### DISCUSSION TITLE
Adding a generic type parameter for using's resource

### DISCUSSION THREAD

**Discussion by jennings** [ReactiveX/rxjs#6497] [Ideas / Feature request]
`using` has the following signature:

```typescript
export function using<T extends ObservableInput<any>>(
  resourceFactory: () => Unsubscribable | void,
  observableFactory: (resource: Unsubscribable  | void) => T | void
): Observable<ObservedValueOf<T>>;
```

This signature means my observableFactory needs to cast its parameter to the actual type:

```typescript
import { using } from "rxjs";

class C {
    unsubscribe() {}
    execute() { return 1 }
}

using(
  () => new C(),
  c => of(c.execute()) // error ts(2339): Property 'execute' does not exist on type 'void | Unsubscribable'.
)
```

Is there any reason not to change the signature like so?

```typescript
export function using<T extends ObservableInput<any>, R extends Unsubscribable = Unsubscribable>(
  resourceFactory: () => R | void,
  observableFactory: (resource: R | void) => T | void
): Observable<ObservedValueOf<T>>;

// and the following dtslint test:
it('should infer resource types', () => {
  const o = using(() => new Subscription(), (sub) => of(sub)); // $ExpectType Observable<Subscription>
});
```

I tried making this change and it seems to work, but there could be a backwards compatibility reason not to change this.

**Comment by cartant** [comment]
I suspect that `using` received close to no attention because it's not something that's used (pardon the pun) much.

It would make sense to use an additional type parameter, but I'd be inclined to go further an add an overload so that the `void` union can be removed, too:

```ts
export function using<T extends ObservableInput<any>, R extends Unsubscribable>(
  resourceFactory: () => R,
  observableFactory: (resource: R) => T | void
): Observable<ObservedValueOf<T>>;

export function using<T extends ObservableInput<any>>(
  resourceFactory: () => void,
  observableFactory: () => T | void
): Observable<ObservedValueOf<T>>;
```

IDK why the `observableFactory` is allowed to return `void` though. That doesn't make much sense. IIRC, `defer` was changed in version 7 so that it _has_ to return an `ObservableInput`. This should probably be changed, too. And maybe allowing the `resourceFactory` to return `void` should be scrapped, too - in which case there'd be no reason for the overload.

Because this changes the type parameters, I don't think this change could be incorporated in version 7, but it could be made in version 8. If you want to open a PR with these changes, please do. Although, it will be labelled as v8 and won't be merged for some time.

**Reply by jennings** [reply]
Thanks, [I've opened a PR](https://github.com/ReactiveX/rxjs/pull/6501) that includes your suggestion.
